specs/codecs/wac/spec.md

```diff
@@ -0,0 +1,192 @@
+* [Implementations](#implementations)
+    * [Go](#go)
+    * [Rust/WASM](#rustwasm)
+
+WAC supports bidirectional transport to and from the [IPLD Data Model].
+In that way it is both a "complete" IPLD Data Model representation and "fitted" to the IPLD Data Model.
+Terminology taken from [Codecs and Completeness].
+
+It takes some inspiration from [Simple DAG], but differs in ways designed to make it reliably round-trip through the IPLD Data Model without loss of fidelity.
+
+The initial motivating use case for WAC is to be a codec that can reliably transport IPLD Data Model data across WebAssembly boundaries while also not being particularly complex to implement.
+
+## Format
+
+### Spec
+
+This format is a series of typing tokens, constant tokens, and inline value data.
+
+Typing tokens are proceeded with the value data for that type.
+
+Every type value can be parsed knowing only the type and without any outside context
+like the container or positional delimiters.
+
+#### Tokens
+
+TODO: Make constants in code match spec
+
+Note: The existing implementations of WAC in https://github.com/aschmahmann/wasm-ipld (Go + Rust/WASM) have different token numbers that we'll need to synchronize before merge.
+
+The tokens for the codec roughly correspond with the [IPLD Data Model Kinds]
+
+They are:
+	Null   = 0
+	True   = 1
+	False  = 2
+	Int    = 3
+	NInt   = 4
+	Float  = 5
+	String = 6
+	Bytes  = 7
+	List   = 8
+	Map    = 9
+	Link   = 10
+
+| Int | Token |
+|---|---|
+| 0 | TYPE_LINK |
+| 1 | TYPE_INTEGER |
+| 2 | TYPE_NEGATIVE_INTEGER |
+| 3 | TYPE_FLOAT |
+| 4 | TYPE_NEGATIVE_FLOAT |
+| 5 | TYPE_STRING |
+| 6 | TYPE_BINARY |
+| 7 | TYPE_MAP |
+| 8 | TYPE_LIST |
+| 9 | VALUE_NULL |
+| 10 | VALUE_TRUE |
+| 11 | VALUE_FALSE |
+
+#### TYPE_LINK
+
+```
+| 0 | CID |
+```
+
+Note: Simple-DAG put a VARINT_LENGTH before the CID indicating how long it was.
+
+CIDs are self-delimiting so this didn't seem necessary.
+
+#### TYPE_INTEGER
+
+```
+| 1 | VARINT |
+```
+
+#### TYPE_NEGATIVE_INTEGER
+
+```
+| 2 | VARINT |
+```
+
+#### TYPE_FLOAT
+
+```
+| 3 | MATISSA_LENGTH | VARINT
+```
+
+TODO: Floats (and negative floats) need definition here.
+Making implementations actually bidirectional with respect to the IPLD Data Model seems difficult here.
+
+#### TYPE_NEGATIVE_FLOAT
+
+```
+| 4 | MATISSA_LENGTH | VARINT
+```
+
+#### TYPE_STRING
+
+```
+| 5 | VARINT_LENGTH | STRING
+```
+
+Note: This is essentially the same as Binary, but with a different token flag.
+
+As with the IPLD Data Model itself Strings are basically just Bytes with something to tag them as "string-like" but where UTF-8 is encouraged. See [Strings].
+
+#### TYPE_BINARY
+
+```
+| 6 | VARINT_LENGTH | BINARY
+```
+
+#### TYPE_MAP
+
+```
+| 7 | VARINT_NUM_PAIRS | PAIRS
+```
+
+`PAIRS` contains alternating keys then values concatenated. i.e. ` KEY1 | VALUE1 | KEY2 | VALUE2 ...`
+
+This codec does not have any form of canonical map sorting as that would make it ill-fitted to the IPLD Data Model.
```

Comment:
It might be worth having the discussion about ordering .. it's been discussed nearly to death but it has some very important things worth acknowledging. This definition of fittedness is an Eric construct; we all line up differently on this topic. There's tradeoffs either way you roll with this.

Reply:
> It might be worth having the discussion about ordering

Yep, I recognize that this PR has sort of kicked the hornets nest in reminding us of all the data model issues we've left unresolved and/or have left some folks unhappy with.

This is happening basically because for a few reasons, like ease of development and performance, in order to enable the WASM work in https://github.com/aschmahmann/wasm-ipld I wanted to be able to work with data model data without creating tons of functions that look like the node building and node iterating functions that exist in go-ipld-prime. This means I've moved some of the pieces of the data model that are "left as an exercise to the implementer" into a spec for a codec we can discuss. It also means that hopefully the wasm-ipld work isn't overly biased towards my perspective coming from the Go libraries. It's definitely biased since it's the only place I've written bindings for and go-ipld-prime is the only library we have ADL support, but I'm trying to minimize the bias and publicly communicate where I can.

> This definition of fittedness is an Eric construct

A bunch of the definitions in that linked document are his construct, but the point remains that if the data model doesn't have a global ordering and WAC wants to be bidirectional with the data model (e.g. I can take an abstract data model node and then serialize and deserialize it and get the same data) then it needs to not care about map ordering as well.

> There's tradeoffs either way you roll with this.

I don't think we run into any serious tradeoffs in this codec around ordering (at least for it's initially intended purpose), however there are some at the data model layer (which this codec is trying to be bidirectional with).

I think https://github.com/ipld/go-ipld-prime/pull/220 was the most recent time this came up. The best resource I can locate at the moment with some of the tradeoffs and analysis here is https://github.com/ipld/ipld/blob/f059543ae6882d0d475a4d0b10ed4bbe1c421fed/design/tricky-choices/ordering.md, but if you've got more to look at (or comments to add) please post them 🙏.

Reply:
_(This is far from exhaustive, it's just top-of-mind while I'm tired and at the end of my day, this is a bit topic)_

https://github.com/ipld/go-ipld-prime/issues/343 is another example of where this comes up, doing this is near top of my list of next things to bite off. With current suite of codecs, the caveats to document there are fewer than when we have a codec that preserves "data model order" (because "data model order" really is a novel concept to us, as https://github.com/ipld/go-ipld-prime/pull/220 demonstrated and https://github.com/ipld/go-ipld-prime/pull/251 is related too). We have similar problems in JS where we're not supposed to depend on object key order (although in practice we can/do). Consider the problem of round-tripping through another codec (dag-json and dag-cbor have different sorting). But the biggest item of concern is convergence/determinism/canonical-forms. When you have a data tightly bound to a schema then you _tend_ toward canonical forms (at the whims of an implementation), when you don't have a schema you're down to initial insertion-order.

Insertion-order preservation as a feature is something that's been argued for before, but that can be done already with List representations, that's what a List is good at.

Maybe it would be interesting to have a non-map-sorting codec in the mix, but you're going to get some hate for it when people bring their ordering expectations to it (maybe it goes the other way today and we don't hear about it? but I see it more in the other way, like when cbor-gen and our other dag-cbor implementations disagree on sorting order and we get different bytes that need explaining—this happens).

Reply:
And for the record, I'm in the mixed-feelings camp on this. I'd love there to be a middle way that allows order preservation to be dictated down from the data model but a default to some kind of sorting without an opinion being expressed by another part of the stack. e.g. Typed (schema-backed) graphs don't get sorting, but untyped (no schema) graphs get sorting. But an argument could be made that this just inserts a whole new set of problems to the mix!